test/EndToEndTests/Tests/Edm/Microsoft.OData.Edm.E2E.Tests/Common/EdmLibTestCaseBase.cs

```diff
@@ -98,6 +98,74 @@ protected static XElement ExtractElementByName(IEnumerable<XElement> inputSchema
         return containers;
     }
 
+    public static XNamespace GetCsdlFullNamespace(EdmVersion csdlVersion)
+    {
+        return EdmStringConstants.EdmOasisNamespace;
+    }
+
+    public static IEnumerable<string> GetBaseTypes(XElement csdl, string entityFullTypeName)
+    {
+        var entityTypeName = entityFullTypeName.Split('.').Last();
+        var entityTypeNamespace = entityFullTypeName.Substring(0, entityFullTypeName.Length - ('.' + entityTypeName).Length);
+
+        var baseTypes = new List<string>();
+        if (csdl.Attribute("Namespace").Value != entityTypeNamespace)
+        {
+            return baseTypes;
+        }
+
+        var types = csdl.Elements(XName.Get("EntityType", csdl.Name.NamespaceName)).Where(n => n.Attribute("Name").Value == entityTypeName);
+        foreach (var type in types)
+        {
+            if (type.Attribute("BaseType") != null)
+            {
+                baseTypes.AddRange(GetBaseTypes(csdl, type.Attribute("BaseType").Value));
+                baseTypes.Add(type.Attribute("BaseType").Value);
+            }
+        }
+        return baseTypes;
+    }
+
+    public static IEnumerable<string> GetDirectlyDerivedTypes(XElement csdl, string structuralTypeElementName, string fullTypeName)
+    {
+        var entityTypeName = fullTypeName.Split('.').Last();
+        var entityTypeNamespace = fullTypeName.Substring(0, fullTypeName.Length - ('.' + entityTypeName).Length);
+
+        var derivedTypes = new List<string>();
+        var types = csdl.Elements(XName.Get(structuralTypeElementName, csdl.Name.NamespaceName)).Where(n => null != n.Attribute("BaseType") && fullTypeName == n.Attribute("BaseType").Value);
+        foreach (var type in types)
+        {
+            if (null != type.Attribute("Name"))
+            {
+                var typeFullName = csdl.Attribute("Namespace").Value + "." + type.Attribute("Name").Value;
+                derivedTypes.Add(typeFullName);
+            }
+        }
+        return derivedTypes;
+    }
+
+    public static IEnumerable<string> GetDirectlyDerivedTypes(IEnumerable<XElement> csdlElements, string structuralTypeElementName, string fullTypeName)
+    {
+        var derivedTypes = new List<string>();
+        foreach (var csdlElement in csdlElements)
+        {
+            derivedTypes.AddRange(GetDirectlyDerivedTypes(csdlElement, structuralTypeElementName, fullTypeName));
+        }

```

Comment:
Dynamically adding items to a list (e.g., with Add or AddRange in a loop) can be costly due to repeated memory allocations and copying. If the final size is known or can be estimated, consider preallocating the list using new List<T>(capacity).
Alternatively, if you're just aggregating and don't need to mutate or index the result, using SelectMany with deferred execution can be more efficient:```csharp IEnumerable<string> derivedTypes = csdlElements
    .SelectMany(csdlElement => GetDirectlyDerivedTypes(csdlElement, structuralTypeElementName, fullTypeName)); ```